File: C++/5-BinarySearch/Algorithms/2-UpperBound&LowerBound/UpperBound&LowerBoundPro.js
// Binary Search: Upper Bound & Lower Bound

/*
    Problem:
    Given a sorted array of integers, find the lower and upper bounds of a target value.
    -   Lower bound: The index of the first occurrence of the target value.
    -   Upper bound: The index of the last occurrence of the target value.

    If the target value is not found, return -1 for both.

    Constraints:
    -   The input array is sorted in ascending order.
    -   The array may contain duplicate elements.
    -   Time complexity should be O(log n).
*/

// Helper function to create test arrays
function createTestArray(size, maxValue, allowDuplicates = true) {
    const arr = [];
    let prev = 0;
    for (let i = 0; i < size; i++) {
        if (allowDuplicates) {
            const increment = Math.floor(Math.random() * (maxValue / size)) + 1;
            prev += increment;
        } else {
            prev += Math.floor(maxValue / size) + 1;
        }
        arr.push(prev);
    }
    return arr;
}

// 1. Iterative Binary Search with Separate Functions

function lowerBoundIterative(arr, target) {
    let low = 0;
    let high = arr.length - 1;
    let result = -1;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] === target) {
            result = mid;       // Found a match, look for earlier one
            high = mid - 1;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return result;
}

function upperBoundIterative(arr, target) {
    let low = 0;
    let high = arr.length - 1;
    let result = -1;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] === target) {
            result = mid;       // Found a match, look for later one
            low = mid + 1;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return result;
}

function binarySearchBoundsIterative(arr, target) {
    const lower = lowerBoundIterative(arr, target);
    const upper = upperBoundIterative(arr, target);
    return [lower, upper];
}



// 2. Iterative Binary Search - Combined Function
function binarySearchBoundsIterativeCombined(arr, target) {
    let lower = -1;
    let upper = -1;
    let low, high, mid;

    // Find Lower Bound
    low = 0;
    high = arr.length - 1;
    while (low <= high) {
        mid = Math.floor((low + high) / 2);
        if (arr[mid] === target) {
            lower = mid;
            high = mid - 1;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    // Find Upper Bound
    low = 0;
    high = arr.length - 1;
    while (low <= high) {
        mid = Math.floor((low + high) / 2);
        if (arr[mid] === target) {
            upper = mid;
            low = mid + 1;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return [lower, upper];
}



// 3. Recursive Binary Search

function binarySearchBoundsRecursive(arr, target, low = 0, high = arr.length - 1, lower = -1, upper = -1, findLower = true) {
    if (low > high) {
        return [lower, upper];
    }

    const mid = Math.floor((low + high) / 2);
    if (arr[mid] === target) {
        if (findLower) {
            lower = mid;
            return binarySearchBoundsRecursive(arr, target, low, mid - 1, lower, upper, true); // Search left for lower
        } else {
            upper = mid;
            return binarySearchBoundsRecursive(arr, target, mid + 1, high, lower, upper, false); // Search right for upper
        }
    } else if (arr[mid] < target) {
        return binarySearchBoundsRecursive(arr, target, mid + 1, high, lower, upper, findLower);
    } else {
        return binarySearchBoundsRecursive(arr, target, low, mid - 1, lower, upper, findLower);
    }
}

function getLowerAndUpperBounds(arr, target) {
    const lower = binarySearchBoundsRecursive(arr, target, 0, arr.length - 1, -1, -1, true)[0];
    const upper = binarySearchBoundsRecursive(arr, target, 0, arr.length - 1, -1, -1, false)[1];
    return [lower, upper];
}



// 4. Using built-in methods (for demonstration - not truly O(log n))
function binarySearchBoundsBuiltIn(arr, target) {
    let lower = arr.indexOf(target);  //  O(n) in worst case
    let upper = arr.lastIndexOf(target); // O(n) in worst case
    return [lower, upper];
}

// 5. Optimized Iterative Approach (Most Efficient)
function binarySearchBoundsOptimized(arr, target) {
    let lower = -1;
    let upper = -1;
    let low = 0;
    let high = arr.length - 1;

    // Find Lower Bound
    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] >= target) { // Changed to >=
            if (arr[mid] === target) lower = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    // Reset for Upper Bound
    low = 0;
    high = arr.length - 1;
    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] <= target) { // Changed to <=
            if (arr[mid] === target) upper = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return [lower, upper];
}



// --- Test Cases ---
const testArray1 = [2, 5, 5, 5, 6, 6, 8, 9, 9, 9];
const target1 = 5;
const target2 = 6;
const target3 = 7;
const target4 = 9;
const target5 = 2;

const testArray2 = createTestArray(20, 100, true); // Allow Duplicates
const target6 = testArray2[Math.floor(testArray2.length / 3)]; // Pick a target within the array
const target7 = 101; // Target not in array

const testArray3 = [7];
const target8 = 7;

const testArray4 = createTestArray(20, 100, false); // No Duplicates
const target9 = testArray4[Math.floor(testArray4.length/2)];

console.log("--- Test Cases ---");
console.log("Test Array 1:", testArray1);
console.log("Target 1 (5):", target1);
console.log("Target 2 (6):", target2);
console.log("Target 3 (7):", target3);
console.log("Target 4 (9):", target4);
console.log("Target 5 (2):", target5);
console.log("Test Array 2 (Duplicates):", testArray2);
console.log("Target 6 (Existing):", target6);
console.log("Target 7 (Not found):", target7);
console.log("Test Array 3 (Single Element):", testArray3);
console.log("Target 8 (Single Element):", target8);
console.log("Test Array 4 (No Duplicates):", testArray4);
console.log("Target 9 (No Duplicates):", target9);


console.log("\n--- Results ---");
console.log("\n1. Iterative with Separate Functions:");
console.log("Test Case 1:", binarySearchBoundsIterative(testArray1, target1)); // [1, 3]
console.log("Test Case 2:", binarySearchBoundsIterative(testArray1, target2)); // [4, 5]
console.log("Test Case 3:", binarySearchBoundsIterative(testArray1, target3)); // [-1, -1]
console.log("Test Case 4:", binarySearchBoundsIterative(testArray1, target4)); // [7, 9]
console.log("Test Case 5:", binarySearchBoundsIterative(testArray1, target5)); // [0, 0]
console.log("Test Case 6 (Random Array):", binarySearchBoundsIterative(testArray2, target6));
console.log("Test Case 7 (Not Found):", binarySearchBoundsIterative(testArray2, target7));
console.log("Test Case 8 (Single Element):", binarySearchBoundsIterative(testArray3, target8));
console.log("Test Case 9 (No Duplicates):", binarySearchBoundsIterative(testArray4, target9));


console.log("\n2. Iterative Combined Function:");
console.log("Test Case 1:", binarySearchBoundsIterativeCombined(testArray1, target1));
console.log("Test Case 2:", binarySearchBoundsIterativeCombined(testArray1, target2));
console.log("Test Case 3:", binarySearchBoundsIterativeCombined(testArray1, target3));
console.log("Test Case 4:", binarySearchBoundsIterativeCombined(testArray1, target4));
console.log("Test Case 5:", binarySearchBoundsIterativeCombined(testArray1, target5));
console.log("Test Case 6:", binarySearchBoundsIterativeCombined(testArray2, target6));
console.log("Test Case 7:", binarySearchBoundsIterativeCombined(testArray2, target7));
console.log("Test Case 8:", binarySearchBoundsIterativeCombined(testArray3, target8));
console.log("Test Case 9:", binarySearchBoundsIterativeCombined(testArray4, target9));


console.log("\n3. Recursive:");
console.log("Test Case 1:", getLowerAndUpperBounds(testArray1, target1));
console.log("Test Case 2:", getLowerAndUpperBounds(testArray1, target2));
console.log("Test Case 3:", getLowerAndUpperBounds(testArray1, target3));
console.log("Test Case 4:", getLowerAndUpperBounds(testArray1, target4));
console.log("Test Case 5:", getLowerAndUpperBounds(testArray1, target5));
console.log("Test Case 6:", getLowerAndUpperBounds(testArray2, target6));
console.log("Test Case 7:", getLowerAndUpperBounds(testArray2, target7));
console.log("Test Case 8:", getLowerAndUpperBounds(testArray3, target8));
console.log("Test Case 9:", getLowerAndUpperBounds(testArray4, target9));


console.log("\n4. Using Built-in Methods (indexOf/lastIndexOf):");
console.log("Test Case 1:", binarySearchBoundsBuiltIn(testArray1, target1));
console.log("Test Case 2:", binarySearchBoundsBuiltIn(testArray1, target2));
console.log("Test Case 3:", binarySearchBoundsBuiltIn(testArray1, target3));
console.log("Test Case 4:", binarySearchBoundsBuiltIn(testArray1, target4));
console.log("Test Case 5:", binarySearchBoundsBuiltIn(testArray1, target5));
console.log("Test Case 6:", binarySearchBoundsBuiltIn(testArray2, target6));
console.log("Test Case 7:", binarySearchBoundsBuiltIn(testArray2, target7));
console.log("Test Case 8:", binarySearchBoundsBuiltIn(testArray3, target8));
console.log("Test Case 9:", binarySearchBoundsBuiltIn(testArray4, target9));

console.log("\n5. Optimized Iterative Approach:");
console.log("Test Case 1:", binarySearchBoundsOptimized(testArray1, target1));
console.log("Test Case 2:", binarySearchBoundsOptimized(testArray1, target2));
console.log("Test Case 3:", binarySearchBoundsOptimized(testArray1, target3));
console.log("Test Case 4:", binarySearchBoundsOptimized(testArray1, target4));
console.log("Test Case 5:", binarySearchBoundsOptimized(testArray1, target5));
console.log("Test Case 6:", binarySearchBoundsOptimized(testArray2, target6));
console.log("Test Case 7:", binarySearchBoundsOptimized(testArray2, target7));
console.log("Test Case 8:", binarySearchBoundsOptimized(testArray3, target8));
console.log("Test Case 9:", binarySearchBoundsOptimized(testArray4, target9));

File: C++/5-BinarySearch/Algorithms/2-UpperBound&LowerBound/UpperBound&LowerBoundPro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>
#include <limits> // Required for numeric_limits

// *************************************************************************************************
// 1. Lower Bound using Standard Library (std::lower_bound)
//
//   - Uses the C++ standard library's `std::lower_bound` function.
//   - Returns an iterator to the first element in the range [first, last) which is
//     not less than (i.e., greater than or equal to) val.
//   - Real-world application: Finding the starting index of a sorted range of products
//     in a database where the price is greater than or equal to a given value.
// *************************************************************************************************
int lowerBoundSTL(const std::vector<int>& arr, int target) {
    auto it = std::lower_bound(arr.begin(), arr.end(), target);
    if (it != arr.end()) {
        return std::distance(arr.begin(), it); // Convert iterator to index
    }
    return -1; // Target not found (or greater than all elements)
}

// *************************************************************************************************
// 2. Upper Bound using Standard Library (std::upper_bound)
//
//   - Uses the C++ standard library's `std::upper_bound` function.
//   - Returns an iterator to the first element in the range [first, last) which is
//     greater than val.
//   - Real-world application: Finding the ending index (exclusive) of a sorted range
//     of products in a database where the price is less than a given value.  Useful
//     for determining the number of products within a price range.
// *************************************************************************************************
int upperBoundSTL(const std::vector<int>& arr, int target) {
    auto it = std::upper_bound(arr.begin(), arr.end(), target);
    if (it != arr.end()) {
        return std::distance(arr.begin(), it);
    }
    return -1; // Target not found (or greater than all elements)
}

// *************************************************************************************************
// 3. Lower Bound (Iterative Binary Search)
//
//   - Implements lower bound using a manual iterative binary search.
//   - This is a fundamental implementation, useful for understanding the algorithm.
//   - Real-world application: Implementing a custom search in a memory-constrained
//     embedded system where using the standard library is not feasible.  Also,
//     good for interview questions!
// *************************************************************************************************
int lowerBoundIterative(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size(); // Important: right is one past the end
    int result = -1;

    while (left < right) {
        int mid = left + (right - left) / 2; // Prevent potential overflow
        if (arr[mid] >= target) {
            result = mid;     // Potential lower bound, try to find a tighter one
            right = mid;     // Search in the left half
        } else {
            left = mid + 1; // Search in the right half
        }
    }
    return result;
}

// *************************************************************************************************
// 4. Upper Bound (Iterative Binary Search)
//
//   - Implements upper bound using a manual iterative binary search.
//   - Mirrors the lower bound implementation but with a slightly different condition.
//   - Real-world application:  Finding the insertion point in a sorted array to
//     maintain sorted order *after* insertion.
// *************************************************************************************************
int upperBoundIterative(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size(); // Important: right is one past the end
    int result = -1;

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] > target) { // Key difference: > instead of >=
            result = mid;
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    return result;
}

// *************************************************************************************************
// 5. Lower Bound with Duplicates Handling (Finds the *first* occurrence)
//
//   - Handles the case where the target value may appear multiple times in the array.
//   - Ensures that the *lowest* index of the target value is returned.
//   - Real-world application: Finding the first occurrence of a specific log entry
//     in a sorted log file, where entries may be duplicated.
// *************************************************************************************************
int lowerBoundFirstOccurrence(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size();
    int result = -1;

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            result = mid;      // Found a match, but check for earlier occurrences
            right = mid;      // Continue searching in the left half
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return result;
}

// *************************************************************************************************
// 6. Upper Bound with Duplicates Handling (Finds the *last* occurrence)
//    This was added as an extra, beyond the original 5.
//
//   - Handles the case where the target value may appear multiple times.
//   - Ensures that the *highest* index of the target value is returned.
//   - Real world application: Find the last entry of a stock trade.
// *************************************************************************************************
int upperBoundLastOccurrence(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size();
    int result = -1;

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            result = mid;
            left = mid + 1; // Look for later occurrences.
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return result; // returns the last index, or -1 if not found
}

// *************************************************************************************************
// 7.  Count occurrences of target (using lower/upper bound)
//      This was added as an extra, beyond the original 5.
//
//   - Uses lower and upper bound to count the number of times a target value
//     appears in the sorted array.
//   - Real-world application: Counting the number of orders for a specific product
//     in an order database.
// *************************************************************************************************
int countOccurrences(const std::vector<int>& arr, int target) {
    int lower = lowerBoundFirstOccurrence(arr, target);
    int upper = upperBoundLastOccurrence(arr, target);
    if (lower == -1) {
        return 0; // Target not found
    }
    return upper - lower + 1;
}

// *************************************************************************************************
// Main function to demonstrate the different implementations
// *************************************************************************************************
int main() {
    std::vector<int> sortedArray = {2, 5, 5, 5, 6, 6, 8, 9, 9, 10};
    int targetValue = 5;

    std::cout << "Sorted Array: ";
    for (int num : sortedArray) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    std::cout << "Target Value: " << targetValue << std::endl;
    std::cout << std::endl;

    // 1. Lower Bound using STL
    int lowerBound1 = lowerBoundSTL(sortedArray, targetValue);
    std::cout << "1. Lower Bound (STL): " << (lowerBound1 != -1 ? std::to_string(lowerBound1) : "Not Found") << std::endl;
    if (lowerBound1 != -1) {
        assert(sortedArray[lowerBound1] >= targetValue);
    }

    // 2. Upper Bound using STL
    int upperBound1 = upperBoundSTL(sortedArray, targetValue);
    std::cout << "2. Upper Bound (STL): " << (upperBound1 != -1 ? std::to_string(upperBound1) : "Not Found") << std::endl;
     if (upperBound1 != -1) {
        assert(sortedArray[upperBound1] > targetValue);
    }

    // 3. Lower Bound (Iterative)
    int lowerBound2 = lowerBoundIterative(sortedArray, targetValue);
    std::cout << "3. Lower Bound (Iterative): " << (lowerBound2 != -1 ? std::to_string(lowerBound2) : "Not Found") << std::endl;
    if (lowerBound2 != -1) {
        assert(sortedArray[lowerBound2] >= targetValue);
    }

    // 4. Upper Bound (Iterative)
    int upperBound2 = upperBoundIterative(sortedArray, targetValue);
    std::cout << "4. Upper Bound (Iterative): " << (upperBound2 != -1 ? std::to_string(upperBound2) : "Not Found") << std::endl;
    if (upperBound2 != -1) {
        assert(sortedArray[upperBound2] > targetValue);
    }
    // 5. Lower Bound First Occurrence
    int lowerBoundFirst = lowerBoundFirstOccurrence(sortedArray, targetValue);
    std::cout << "5. Lower Bound (First Occurrence): " << (lowerBoundFirst != -1 ? std::to_string(lowerBoundFirst) : "Not Found") << std::endl;
    if (lowerBoundFirst != -1) {
         assert(sortedArray[lowerBoundFirst] == targetValue);
         if (lowerBoundFirst > 0) {
            assert(sortedArray[lowerBoundFirst-1] < targetValue);
         }
    }

    // 6. Upper Bound Last Occurrence
    int upperBoundLast = upperBoundLastOccurrence(sortedArray, targetValue);
    std::cout << "6. Upper Bound (Last Occurrence): " << (upperBoundLast != -1 ? std::to_string(upperBoundLast) : "Not Found") << std::endl;
    if (upperBoundLast != -1){
        assert(sortedArray[upperBoundLast] == targetValue);
        if(upperBoundLast < sortedArray.size() - 1){
             assert(sortedArray[upperBoundLast + 1] > targetValue);
        }

    }
    // 7. Count Occurrences
    int count = countOccurrences(sortedArray, targetValue);
    std::cout << "7. Count Occurrences: " << count << std::endl;
    if (count > 0) {
        assert(lowerBoundFirstOccurrence(sortedArray, targetValue) != -1);
        assert(upperBoundLastOccurrence(sortedArray, targetValue) != -1);
    }

    // Test case where the target is not in the array
    targetValue = 7;
    std::cout << "\nTesting with target value " << targetValue << std::endl;
    std::cout << "1. Lower Bound (STL): " << (lowerBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "2. Upper Bound (STL): " << (upperBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(upperBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "3. Lower Bound (Iterative): " << (lowerBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "4. Upper Bound (Iterative): " << (upperBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(upperBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "5. Lower Bound (First Occurrence): " << (lowerBoundFirstOccurrence(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundFirstOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "6. Upper Bound (Last Occurrence): " << (upperBoundLastOccurrence(sortedArray, targetValue) != -1 ? std::to_string(upperBoundLastOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "7. Count Occurrences: " << countOccurrences(sortedArray, targetValue) << std::endl;

    // Test case where the target is smaller than all elements
    targetValue = 1;
    std::cout << "\nTesting with target value " << targetValue << std::endl;
    std::cout << "1. Lower Bound (STL): " << (lowerBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "2. Upper Bound (STL): " << (upperBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(upperBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "3. Lower Bound (Iterative): " << (lowerBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "4. Upper Bound (Iterative): " << (upperBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(upperBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "5. Lower Bound (First Occurrence): " << (lowerBoundFirstOccurrence(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundFirstOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "6. Upper Bound (Last Occurrence): " << (upperBoundLastOccurrence(sortedArray, targetValue) != -1 ? std::to_string(upperBoundLastOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "7. Count Occurrences: " << countOccurrences(sortedArray, targetValue) << std::endl;

    // Test case where the target is larger than all elements
    targetValue = 11;
    std::cout << "\nTesting with target value " << targetValue << std::endl;
    std::cout << "1. Lower Bound (STL): " << (lowerBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "2. Upper Bound (STL): " << (upperBoundSTL(sortedArray, targetValue) != -1 ? std::to_string(upperBoundSTL(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "3. Lower Bound (Iterative): " << (lowerBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "4. Upper Bound (Iterative): " << (upperBoundIterative(sortedArray, targetValue) != -1 ? std::to_string(upperBoundIterative(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "5. Lower Bound (First Occurrence): " << (lowerBoundFirstOccurrence(sortedArray, targetValue) != -1 ? std::to_string(lowerBoundFirstOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "6. Upper Bound (Last Occurrence): " << (upperBoundLastOccurrence(sortedArray, targetValue) != -1 ? std::to_string(upperBoundLastOccurrence(sortedArray, targetValue)) : "Not Found") << std::endl;
    std::cout << "7. Count Occurrences: " << countOccurrences(sortedArray, targetValue) << std::endl;

    return 0;
}

